### PR TITLE
doc: add RFC for the sysprof

### PR DESCRIPTION
**Pull request by fckxorg**
This commit adds RFC for the sysprof that was approved a year ago.

Part of #781

NO_TEST=RFC
NO_CHANGELOG=RFC

### CODE REVIEW

File: doc/rfc/781-lua-system-wide-profiler.md [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+The document describes the platform profiler for LuaJIT. It is needed to obtain a complete view of platform performance. The existing LuaJIT profiler is only able to give you information about virtual machine states and guest stacks. Hence, the document proposes to provide a new sampling profiler, so it will be able to gather stack traces from both C and Lua.
+
+## Background and motivation
+
+Currently, the available options for profiling LuaJIT are not fine enough to provide an understanding of performance. For example, perf is only able to show the host stack, so all the Lua calls are seen as a single pcall. Oppositely, the jit.p module provided with LuaJIT is not able to give any information about the host stack.
+
+To get a detailed perspective on the platform's performance, a more advanced profiler is needed. The desired profiler must be able to capture both guest and host stacks simultaneously, along with virtual machine states.
+
+Assume we have the following chunks of code:
+```c=
```
Comment (Buristan) [inline doc/rfc/781-lua-system-wide-profiler.md:18]: Unfortunately, GitHub doesn't support `=` for line numbers, so syntax highlighting is broken. You can add those line manually for Lua code to be more readable (when we really need line numbers). Here and below.

File: doc/rfc/781-lua-system-wide-profiler.md [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+
+| ![](https://i.imgur.com/X0nL83r.png) |
+| :--: |
+| `CALLCHAIN` mode |
+
+
+The `sampling_interval` is the sampling period in milliseconds. `path` is name of the file where profile events are written. This function's writer calls `fwrite()` for each call, retrying if EINTR occurs. When the profiling is stopped, `fclose()` is called. The contract for the profiler function is similar to that of standard io* interfaces. If it is impossible to open a file for writing or the profiler fails to start, nil is returned on failure (plus an error message as a second result and a system-dependent error code as a third result). Otherwise, it returns the `true` value.
+
+Stopping profiler from Lua is simple too:
+```
```
Comment (Buristan) [inline doc/rfc/781-lua-system-wide-profiler.md:361]: It will be nice to add Lua syntax highlighting here too.

**Review by Buristan — APPROVED** [review]
Hi, Maxim!
Thanks for the patch!
LGTM with 2 comments below.

File: doc/rfc/781-lua-system-wide-profiler.md [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+
+```lua
+ 1  -- file lua_payload.lua
+ 2
+ 3  local jit = require("jit")
+ 4  local ffi = require("ffi")
+ 5  local os = require("os")
+ 6
+ 7  ------------------------------------------------
+ 8
```
Comment (Buristan) [inline doc/rfc/781-lua-system-wide-profiler.md:38]: I noticed, that we have `@../lua_lib.lua:8` on flamegraph, but this it should be `:10` shouldn't it?

File: doc/rfc/781-lua-system-wide-profiler.md [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+As you can see, the writer backend is customizable, and our implementation does not lock it at the C-API level.
+It is required to provide a writer, a backtracer, and an on-stop callback before running the profiler. If one wants to use the default backtracer, he can call `luaM_sysprof_set_backtracer` with `NULL` argument.
+To start the profiler, one needs to fill in the `luam_Sysprof_Options` with desired values and pass it to the `luaM_sysprof_start` function. In the case of a successful start, it returns `PROFILE_SUCCESS`.
+
+
+To stop the profiler, one should call `luaM_sysprof_stop` function, which will return `PROFILE_SUCCESS` in case of successful profiling finalization.
+
+Starting the profiler in Lua is simple:
+
+```lua
```
Comment (Buristan) [inline doc/rfc/781-lua-system-wide-profiler.md:326]: As far as we don't reference this code we can omit line numbers.
Feel free to ignore.

File: doc/rfc/781-lua-system-wide-profiler.md [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+| ![](https://i.imgur.com/X0nL83r.png) |
+| :--: |
+| `CALLCHAIN` mode |
+
+
+The `sampling_interval` is the sampling period in milliseconds. `path` is name of the file where profile events are written. This function's writer calls `fwrite()` for each call, retrying if EINTR occurs. When the profiling is stopped, `fclose()` is called. The contract for the profiler function is similar to that of standard io* interfaces. If it is impossible to open a file for writing or the profiler fails to start, nil is returned on failure (plus an error message as a second result and a system-dependent error code as a third result). Otherwise, it returns the `true` value.
+
+Stopping profiler from Lua is simple too:
+```lua
+1  local stopped, err, errno = misc.sysprof.stop()
```
Comment (Buristan) [inline doc/rfc/781-lua-system-wide-profiler.md:362]: As far as we don't reference this code we can omit line numbers.
Feel free to ignore.

**Review by Buristan — APPROVED** [review]
Hi, Maxim!
Thanks for the fixes!
LGTM, except 2 nits and one question.

File: doc/rfc/781-lua-system-wide-profiler.md (hunk 3 above)
Comment (igormunkin) [reply doc/rfc/781-lua-system-wide-profiler.md:38]: I have no strict opinion for this: there would be nice to use the actual flamegraph here, but I believe we can ignore this (or fix on demand if someone is confused by it). If everybody is OK, I'll simply merge this PR as is.

**Review by igormunkin — APPROVED** [review]
@fckxorg, thanks for the patch! Looks like the one you've implemented, so LGTM :)

File: doc/rfc/781-lua-system-wide-profiler.md [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+39      i = i + 1
+40  end
+```
+
+The `c_payload.c` code was compiled into a shared library and loaded by LuaJIT via the FFI.
+
+To see the difference, you can take a look at the flamegraphs (learn more about [flamegraphs](https://www.brendangregg.com/flamegraphs.html)) for this code generated by perf, jit.p, and PoC for the proposed profiler below.
+
+### perf
+![perf](https://i.imgur.com/DlKbFpo.png)
```
Comment (Totktonada) [inline doc/rfc/781-lua-system-wide-profiler.md:78]: Third party services may gone. I suggest to add a directory and place the images right here. You can use a relative URI: `perf`.

**Review by Buristan — APPROVED** [review]
LGTM, after fixes with image sources.

**Pull request merged**